### ISSUE TITLE
Support "negating" style in naming conventions for EditorConfig

### ISSUE THREAD

**Issue by kuhlenh** [dotnet/roslyn#18354]
Today, I can write a naming convention that says:

> Enforce names to start with "xyz" prefix / end with "abc" suffix / etc.

However, sometimes I want to prohibit the use of a prefix/suffix/etc. to have a rule that is more like:

> Prevent field names from starting with "m_" or "_"

**Comment by Pilchie** [comment]
How is this different than not specifying a prefix or word separator and picking a capitilization story?

**Comment by kuhlenh** [comment]
After talking with @Pilchie, this is "by design". A team should be able to configure the rule as
> Field names must be lowercase

and if the word separator is empty, then any occurrence of an underscore will trigger a violation. This is being tracked here now: https://github.com/dotnet/roslyn/issues/18409

**Comment by sharwell** [comment]
I was just looking at how the naming styles configuration might relate to #18467, and realized that while it's possible to enforce that a particular method ends in `Async`, it's not possible to enforce that a method does *not* end with this suffix. Given that there is also no way to create a naming specification based on the return value, it's probably not the only limiting factor for creating an "Avoid Async suffix" convention, but it still seemed like the lack of negation would prove to be one limiting factor here.

**Comment by xperiandri** [comment]
Yep, this must be implemented! Just a single negating field would be nice.

**Comment by xperiandri** [comment]
So when? Critical feature!

**Comment by damiarnold** [comment]
I was just looking for the exact same thing mentioned by @sharwell when I stumbled upon this thread.... the ability to specify that non-async methods do *not* end in "Async".  
Here's hoping this feature gets included soon!

**Comment by xperiandri** [comment]
Moreover if your async method is an MVC controller method it must not end with Async!
How to handle this case too?

**Comment by jnm2** [comment]
Non-root .editorconfig in the controllers folder which disables the rule?

**Comment by xperiandri** [comment]
However if controller have private async utility method then it must have Async suffix.
So the rule is more complicated

**Comment by jnm2** [comment]
At the end of the day, .editorconfig can't know about framework-specific rules like those that apply to MVC controller methods. You might have to err on the loose side with the .editorconfig at some stage, What would be cool is it was possible for an MVC analyzer to suppress .editorconfig examination of controller methods, applying its own diagnostics.

**Comment by jnm2** [comment]
At the end of the day, .editorconfig can't know about framework-specific rules like those that apply to MVC controller methods. You might have to err on the loose side with the .editorconfig at some stage.

What would be cool is if it was possible for an MVC analyzer to suppress .editorconfig examination of controller methods, applying its own diagnostics.

**Comment by hajekj** [comment]
Or maybe add some sort of exception if it is inherited from some class like `Microsoft.AspNetCore.Mvc.Controller`?

**Comment by arfmach** [comment]
> I was just looking for the exact same thing mentioned by @sharwell when I stumbled upon this thread.... the ability to specify that non-async methods do _not_ end in "Async".
> Here's hoping this feature gets included soon!

One must not forget that it implies two negating rule types:
1. method name do _not_ end in "Async" (negate `required_suffix`); and
2. method has _not_ the modifier "async" (negate `required_modifiers`)

**Comment by CyrusNajmabadi** [comment]
> method has not the modifier "async" (negate required_modifiers)

This is not a naming style.  This is a personal repo choice about it async mtehods are allowed or not.  I don't think the Roslyn IDE should have any styles that restrict that sort of thing.  If you want to prevent async-methods in your codebase, just write your own analyzers that looks for this and errors if it finds it.

**Comment by arfmach** [comment]
@CyrusNajmabadi : no, I think you didn't get the point. It is not about forbidding async methods; as stated above, you can place a rule that says that async methods must have

**Comment by CyrusNajmabadi** [comment]
I don't understand this part then:

> method has not the modifier "async"

There shouldn't be any naming that is dependent on the 'async' modifier.  The 'async' modifier affects the internal parsing of a method.  It has no impact on its external shape, and should not be part of hte naming.  

For example, if i have this real method:

```c#
public Task<int> GetCountAsync()
{
     // some work
    return ComputeCountAsync();
}
```

This method should end with 'Async' even though it has no 'async' modifier, and even though it does not use 'await' inside of it.

**Comment by CyrusNajmabadi** [comment]
> I just said that, to declare the opposite rule (non-async methods must not have the suffix Async)

Methods are asynchronous or not depending on the return type of the method *not* depending on the modifiers of a method.  Negating modifiers doesn't make sense in this case.

**Comment by IanKemp** [comment]
Would be nice to have this to cover scenarios like the following: https://stackoverflow.com/questions/54685182/editorconfig-rules-for-differing-naming-of-private-constants-vs-private-fields

**Comment by sharwell** [comment]
@IanKemp You need to break up the rule for `const` and `readonly`. The `required_modifiers` property is a "match all" constraint. Placing more specific rules first is a standard approach to the naming rules. For example, this repository has a rule which applies to all fields that are _not static_, and we apply the rule by simply defining the rule for static fields first:

https://github.com/dotnet/roslyn/blob/b4a4a984a5d11b079940beb87d9f032ad0250eb3/.editorconfig#L76-L95

**Comment by rubberduck203** [comment]
There’s a lot of talk about async here, but I’d like to propose a much simpler use case for this. 

The best I can tell, the naming rules explicitly ignore leading underscores when specifying `camel_case`, so there’s currently no way to specify that private fields should not have a leading underscore.

If we could negate a prefix, that would provide a solution for  #22745

**Comment by alexshikov** [comment]
What are the chances this feature will move forward?

**Comment by VincentH-Net** [comment]
This feature would be very useful to modernize conventions in existing codebases.
E.g.
- No _ prefix for private fields
- No Async suffix for async methods (unless there also is a non-async overload)

Any chance this will happen? Or an alternative way to do this?

**Comment by javiergacrich** [comment]
any news on this?

**Comment by CyrusNajmabadi** [comment]
> any news on this?

Nope! Sorry! :)

**Comment by smaudet** [comment]
This would be *super* helpful to have. This can't be by "design", that's wrong:

```private int _aMemberField;```
```private const int A_CONST```;

It could just be that the state of editorconfig parsing is broken like a bat outa' hell, but if I understand the parsing rules to do this you would need:

```
# constant style
dotnet_naming_rule.constants_should_be_upper_case.symbols = constants
dotnet_naming_rule.constants_should_be_upper_case.style = constant_style

dotnet_naming_symbols.constants.applicable_kinds = field
dotnet_naming_symbols.constants.required_modifiers = const

dotnet_naming_style.constant_style.capitalization = upper_case
dotnet_naming_style.constant_style.word_separator = _
dotnet_naming_style.constant_style.not_a_prefix = _

# private field style
dotnet_naming_symbols.psyms.applicable_kinds = field
dotnet_naming_symbols.psyms.applicable_accessibilities = private

dotnet_naming_style.prefix_private_camel_case.capitalization = camel_case
dotnet_naming_style.prefix_private_camel_case.required_prefix = _

dotnet_naming_rule.private_camelcase.symbols = psyms
dotnet_naming_rule.private_camelcase.style = prefix_private_camel_case
```

If you **dont** do that, its ambiguous what format `private const` should be.

**Comment by Liam-at-ENSEK** [comment]
This works for us, no more underscores :)

`
# Use PascalCase for constant fields  
dotnet_naming_style.pascal_case_style.capitalization             = pascal_case
dotnet_naming_style.pascal_case_style.not_a_prefix             = _

dotnet_naming_rule.constant_fields_should_be_pascal_case.severity = error
dotnet_naming_rule.constant_fields_should_be_pascal_case.symbols  = constant_fields
dotnet_naming_rule.constant_fields_should_be_pascal_case.style    = pascal_case_style

dotnet_naming_symbols.constant_fields.applicable_kinds            = field
dotnet_naming_symbols.constant_fields.applicable_accessibilities  = *
dotnet_naming_symbols.constant_fields.required_modifiers          = const

# Use camel without _ for private fields
dotnet_naming_style.camel_case_style.capitalization = camel_case
dotnet_naming_style.camel_case_style.not_a_prefix = _

dotnet_naming_rule.instance_fields_should_be_camel_case.severity = error
dotnet_naming_rule.instance_fields_should_be_camel_case.symbols  = instance_fields
dotnet_naming_rule.instance_fields_should_be_camel_case.style    = camel_case_style

dotnet_naming_symbols.instance_fields.applicable_kinds            = field
dotnet_naming_symbols.constant_fields.applicable_accessibilities  = *
dotnet_naming_symbols.constant_fields.required_modifiers          = private
`

**Comment by Liam-at-ENSEK** [comment]
This works for me:

```
# Use camel without _ for private fields
dotnet_naming_style.camel_case_style.capitalization = camel_case
dotnet_naming_style.camel_case_style.not_a_prefix = _

dotnet_naming_rule.instance_fields_should_be_camel_case.severity = error
dotnet_naming_rule.instance_fields_should_be_camel_case.symbols  = instance_fields
dotnet_naming_rule.instance_fields_should_be_camel_case.style    = camel_case_style

dotnet_naming_symbols.instance_fields.applicable_kinds            = field
dotnet_naming_symbols.instance_fields.applicable_accessibilities  = private
dotnet_naming_symbols.instance_fields.required_modifiers          = *
```

**Comment by VincentH-Net** [comment]
I finally found out how to enforce NO underscore prefix for private fields in .editorconfig:
![image](https://user-images.githubusercontent.com/1872271/177199653-89b1bd6c-3f35-4655-8d04-f4a54d19f052.png)

[Tweet](https://twitter.com/vincenth_net/status/1544012069577658369) [Gist](https://gist.github.com/VincentH-Net/a88840c96f0083af3064d74e052ad20c)

NJoy!

**Comment by xperiandri** [comment]
@VincentH-Net thank you very much for your discovery!!!

**Comment by HaberkornJonas** [comment]
In our company, we usually use the prefix "api" on data contract classes used for exchanges with the front-end *(e.g. "apiUser")*
That's the only prefix we allow to be camel case.
All other classes must be pascal case.

I was looking for a way to create an exception for this specific prefix but reading this *(quite old)* thread I understood it's not possible and won't be in the near future.

To go around this limitation, I created a new section in the .editorconfig to target only files starting with "api" where all classes need to start with this prefix, classes in the rest of the files need to be pascal case.

```

[api*.{cs,vb}]
# Internal Api Classes Naming Convention #

dotnet_naming_rule.api_should_start_wit_api.severity = error
dotnet_naming_rule.api_should_start_wit_api.symbols = class_to_prefix
dotnet_naming_rule.api_should_start_wit_api.style = start_with_api

dotnet_naming_style.start_with_api.required_prefix = api
dotnet_naming_style.start_with_api.required_suffix = 
dotnet_naming_style.start_with_api.word_separator = _
dotnet_naming_style.start_with_api.capitalization = pascal_case

dotnet_naming_symbols.class_to_prefix.applicable_kinds = class
dotnet_naming_symbols.class_to_prefix.applicable_accessibilities = *
dotnet_naming_symbols.class_to_prefix.required_modifiers = 


[*.{cs,vb}]
#### Naming styles ####
...
```


I know this is quite specific to our needs but I hope sharing it gives some ideas to some other struggling with this limitation.

**Comment by codeabuse** [comment]
I want my private instance fields to be `_myPrivate`, while private const/readonly to be `my_private_const`, yet the current rules are mixing them up for private constants, requiring them to be `_my_private_const`, which is annoying as hell. Why can't we have a negative expression for categories like `[v] field`, `[x] const`, `[x] readonly`? Or at least let us choose between instance and static members with an addition al `[v] instance` parameter.
